### PR TITLE
refactor: Project cell gradients via vjp of `frame.vectors` instead of `from_matrix`

### LINKED ISSUE #158: Torch MLFFs give incorrect lattice gradients if the frame is not linear.
### What broke?

Currently the torch interface for mlffs calls Frame.from_matrix which works well for linear maps that simply select matrix elements. For non-linear definitions like a logarithmic representation, it would need to be the vjp of the from_matrix function.

### How to reproduce

```python

```

### Expected behavior

_No response_

### Environment (optional)

_No response_

### PR DESCRIPTION
**Pull request by n-gao**
Replace `from_matrix` class method with VJP-based gradient projection for cell frame updates

The `from_matrix` class method was being used in two places to map raw `∂E/∂h` matrices back onto a cell frame's parameter space (e.g., `tril` for triclinic, `lengths` for orthogonal). This approach was incorrect because `from_matrix` recomputes derived quantities like inverse and volume, which is inappropriate when handling gradients — only the relevant degrees of freedom should be populated.

A new helper, `_project_grad_onto_frame`, replaces this pattern in the MLIAP interface. It uses `jax.vjp` on the frame's `vectors` property to correctly pull the raw matrix gradient back onto the frame's own parameter space, preserving the frame type without triggering any unnecessary recomputation.

The same misuse in `CellPositionStep` is also corrected by switching from `.set(TriclinicFrame.from_matrix(...))` to `.apply(lambda x: x.from_matrix(...))`, removing the hard-coded `TriclinicFrame` import and making the call polymorphic over the actual frame type.

Tests are added to verify that both triclinic and orthogonal frames are preserved through the projection and that only the correct components (lower triangle and diagonal, respectively) are populated in the output.

Closes #158

### CODE REVIEW

**Comment by n-gao** [comment]
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#178](https://github.com/cusp-ai-oss/kUPS/pull/178) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#169** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#166** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#165** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/153?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#152** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#151** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#150** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#148** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#147** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/146?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#145** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#144** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#143** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#142** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#141** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#138** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#137** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#136** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/136?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#135** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#122** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#134** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/134?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/133?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#132** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/132?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#131** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/md/integrators.py [hunk 1]
```diff
@@ -1085,8 +1084,8 @@ def __call__(self, key: Array, state: State) -> State:
         # V_old, V_dot both lower-tri ⇒ V_new lower-tri.
         V_new = V_old + (params.time_step[..., None, None] / 2) * V_dot
         # Reconstruct TriclinicFrame so cached volume refreshes.
-        return self.systems.focus(lambda x: x.data.cell.frame).set(
-            state, TriclinicFrame.from_matrix(V_new)
+        return self.systems.focus(lambda x: x.data.cell.frame).apply(
+            state, lambda x: x.from_matrix(V_new)
```
Comment (jonkhler) [inline src/kups/md/integrators.py]: elegant solution, but `x` is a bit confusing as a name in this context

**Review by jonkhler — CHANGES_REQUESTED** [review]
can we find a better naming for those lambdas?

**Comment by n-gao** [comment]
Renamed to `frame`

File: src/kups/md/integrators.py (hunk 1 above)
Comment (n-gao) [reply src/kups/md/integrators.py]: renamed to `frame`

**Review by jonkhler — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Jun 16, 3:44 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166).
